src/client/gpm_common.c

```diff
@@ -300,55 +299,41 @@ static int gpm_epoll_wait(struct gpm_ctx *gpmctx, uint32_t event_flags)
     if (epoll_ret < 0) {
         /* Error while waiting that isn't EINTR */
         ret = errno;
-        gpm_epoll_close(gpmctx);
+        goto done;
     } else if (epoll_ret == 0) {
         ret = ETIMEDOUT;
-        gpm_epoll_close(gpmctx);
-    } else if (epoll_ret == 1 && events[0].data.fd == gpmctx->timerfd) {
-        /* Got an event which is only our timer */
-        if ((events[0].events & EPOLLIN) == 0) {
-            /* We got an event which was not EPOLLIN; assume this is an error,
-             * and exit with EBADF: epoll_wait said timerfd had an event,
-             * but that event is not an EPOLIN event. */
-            ret = EBADF;
-        } else {
-            ret = read(gpmctx->timerfd, &timer_read, sizeof(uint64_t));
-            if (ret == -1 && errno != EAGAIN && errno != EWOULDBLOCK) {
-                /* In the case when reading from the timer failed, don't hide the
-                 * timer error behind ETIMEDOUT such that it isn't retried */
-                ret = errno;
-            } else {
-                /* If ret == 0, then we definitely timed out. Else, if ret == -1
-                 * and errno == EAGAIN or errno == EWOULDBLOCK, we're in a weird
-                 * edge case where epoll thinks the timer can be read, but it
-                 * is blocking more; treat it like a TIMEOUT and retry, as
-                 * nothing around us would handle EAGAIN from timer and retry
-                 * it. */
-                ret = ETIMEDOUT;
-            }
-        }
-        gpm_epoll_close(gpmctx);
-    } else {
-        /* If ret == 2, then we ignore the timerfd; that way if the next
-         * operation cannot be performed immediately, we timeout and retry.
-         * Always check the returned event of the socket fd. */
-        int fd_index = 0;
-        if (epoll_ret == 2 && events[fd_index].data.fd != gpmctx->fd) {
-            fd_index = 1;
+        goto done;
+    } else if (epoll_ret == 2 || events[0].data.fd == gpmctx->timerfd ||
```

Comment:
Same here. Also, this conditional definitely is not correct. It should be:

```c

if (events[0].data.fd == gpmctx->timerfd || (epoll_ret == 2 && events[1].data.fd == gpmctx.timerfd) {
    ...
}
```

I believe. I guess it doesn't really matter with the `memset` on lines 277-279 (and this is exactly what they're meant to prevent... :-) ) but _technically_ you could be reading events that were never initialized when:

 - `epoll_ret == 1` and
 - `events[0].data.fd == gpmctx->fd`.

With the `memset` (and a comment about why they're necessary), you could change it to:

```c
if (events[0].data.fd == gpmctx->timerfd || events[1].data.fd == gpmctx.timerfd) {
    ...
}
```

but I'd avoid that.